[core] Omnibus: Split RenderStyle from Style

jfirebaugh opened the pull request:
This omnibus PR contains the last prerequisites for splitting `RenderStyle` and `Style` (following up #9010), and then the actual split.

The prerequisites consist of refactoring `RenderSource` updates so that they do not depend on `Style` itself, but rather a vector of layers and the results of diffing those layers to the previous set:

* Eliminate `Style::updateBatch` in favor of diffing layers and detecting changes to properties upon which layout depends.
* Replace `RenderSource::{update,remove,invalidate,reload}Tiles` with a single update method (fixes #7663)
* Replace `TilePyramid::{update,remove,invalidate,reload}Tiles` with a single update method
* Remove `Style&` dependency TODO from `GeometryTile` and `TileParameters`

ivovandongen commented:
Test failure seems to be a problem with the test. `TileObserver::onTileError` is never called because the `GeometryTileWorker` [bails](https://github.com/mapbox/mapbox-gl-native/blob/8211acc7d1382d46cf45c6741773baed099a23f0/src/mbgl/tile/geometry_tile_worker.cpp#L267) in `redoLayout` because the layers are not set. They are not set, because they are [retrieved](https://github.com/mapbox/mapbox-gl-native/blob/8211acc7d1382d46cf45c6741773baed099a23f0/src/mbgl/tile/geometry_tile.cpp#L107) from the render style now, but those are never set in this test (would have been done in `Map::Impl::render`.

jfirebaugh commented:
I converted this to an omnibus PR encompassing https://github.com/mapbox/mapbox-gl-native/pull/9143 and https://github.com/mapbox/mapbox-gl-native/pull/9010#discussion_r119127778 as well as the original changes here, and updated the description above. Ready for re-review!

ivovandongen commented:
I'm getting a crash when running the ui tests for Android:

```
********** Crash dump: **********
Build fingerprint: 'google/sailfish/sailfish:7.1.2/N2G47O/3852959:user/release-keys'
pid: 20981, tid: 20981, name: pboxsdk.testapp  >>> com.mapbox.mapboxsdk.testapp <<<
signal 11 (SIGSEGV), code 1 (SEGV_MAPERR), fault addr 0x14
Stack frame #00 pc 001d1798  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine std::__ndk1::shared_ptr<mbgl::style::Layer::Impl const>::get() const at /Users/ivo/Library/Android/sdk/ndk-bundle/sources/cxx-stl/llvm-libc++/include/memory:4041
Stack frame #01 pc 002300cd  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine mbgl::style::Image::getImage() const at /Users/ivo/git/mapbox-gl-native/src/mbgl/style/image.cpp:22
Stack frame #02 pc 00314e91  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine mbgl::AnnotationManager::getTopOffsetPixelsForImage(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&) at /Users/ivo/git/mapbox-gl-native/src/mbgl/annotation/annotation_manager.cpp:220 (discriminator 5)
Stack frame #03 pc 0019a89b  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine mbgl::Map::getTopOffsetPixelsForAnnotationImage(std::__ndk1::basic_string<char, std::__ndk1::char_traits<char>, std::__ndk1::allocator<char> > const&) at /Users/ivo/git/mapbox-gl-native/src/mbgl/map/map.cpp:816 (discriminator 1)
Stack frame #04 pc 000a8781  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine mbgl::android::NativeMapView::getTopOffsetPixelsForAnnotationSymbol(_JNIEnv&, jni::Object<jni::StringTag>) at /Users/ivo/git/mapbox-gl-native/platform/android/src/native_map_view.cpp:743
Stack frame #05 pc 000cf713  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine operator() at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:245 (discriminator 5)
Stack frame #06 pc 000cf699  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine operator() at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:112 (discriminator 2)
Stack frame #07 pc 000cf64b  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine __invoke at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:110
Stack frame #08 pc 000cf7c5  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine operator()<jni::jobject *, jni::jstring *> at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:59
Stack frame #09 pc 000cf77b  /data/app/com.mapbox.mapboxsdk.testapp-1/lib/arm/libmapbox-gl.so: Routine __invoke<jni::jobject *, jni::jstring *> at /Users/ivo/git/mapbox-gl-native/platform/android/MapboxGLAndroidSDK/.externalNativeBuild/cmake/debug/armeabi-v7a/../../../../../../../mason_packages/headers/jni.hpp/3.0.0/include/jni/native_method.hpp:55
Stack frame #10 pc 0054e725  /data/app/com.mapbox.mapboxsdk.testapp-1/oat/arm/base.odex (offset 0x525000)
```

ivovandongen commented:
@jfirebaugh Verified that your changes resolves the crash in `AnnotationManager::getTopOffsetPixelsForImage`.

kkaefer commented on src/mbgl/style/collection.hpp:
```diff
@@ -0,0 +1,140 @@
+}
+
+} // namespace style
+} // namespace mbgl
```
Overall, I'm a little worried that we're doing too many linear searches when doing any sort of operation. Why do we need all of them?

jfirebaugh replied:
We need an ordered collection for layers, and all the collections are small enough that linear search is probably fast enough. I haven't seen it show up in profiling results, and (with the exception of images) this PR isn't really changing the complexity of any operations.

kkaefer commented on src/mbgl/renderer/render_style.cpp:
```diff
@@ -0,0 +1,448 @@
+    for (const auto& entry : renderLayers) {
+        RenderLayer& layer = *entry.second;
+        const bool layerAdded = layerDiff.added.count(entry.first);
+        const bool layerChanged = layerDiff.changed.count(entry.first);
```
We're doing two linear searches for every layer. In the three loops above, we already know whether a layer was added/changed/removed. Can we fold this loop into the other loops? (Same for `hasLayoutDifference` for sources below)

jfirebaugh replied:
`StyleDifference` uses `unordered_map` for `added`/`removed`/`changed` (order isn't important in this particular case), so these are constant time operations. There is string hashing involved, but it doesn't show up as significant in profiling.

kkaefer commented:
@jfirebaugh It's becoming increasingly difficult to understand the architecture of the whole Render* and *Impl system with Immutable wrappers when looking at the code base without having followed along the gradual modifications. I think we need a document, or more inline code comments, describing the general architecture of the system with ownership, and thread association.

ivovandongen commented:
@jfirebaugh Last crash in the Android ui tests was unrelated and solved in https://github.com/mapbox/mapbox-gl-native/pull/9162. All looks good now.

jfirebaugh commented on src/mbgl/style/collection.hpp:
```diff
@@ -0,0 +1,140 @@
+std::size_t Collection<T>::index(const std::string& id) const {
+    return std::find_if(wrappers.begin(), wrappers.end(), [&](const auto& e) {
+        return e->getID() == id;
+    }) - wrappers.begin();
```
It returns `end - begin`, i.e. the collection size. This is checked by callers where needed. The only callers are in `Collection` itself so I'll make it private.

jfirebaugh commented on src/mbgl/style/collection.hpp:
```diff
@@ -0,0 +1,140 @@
+template <class T>
+void Collection<T>::update(const T& wrapper) {
+    mutate(impls, [&] (auto& impls_) {
+        impls_.at(this->index(wrapper.getID())) = wrapper.baseImpl;
```
Since `index` returns the collection size in that case, `at` will throw `std::out_of_range`.

jfirebaugh commented on src/mbgl/renderer/sources/render_raster_source.cpp:
```diff
@@ -44,24 +36,36 @@ void RenderRasterSource::updateTiles(const TileParameters& parameters) {
-        tilePyramid.invalidateTiles();
+
+        // TODO: this removes existing buckets, and will cause flickering.
+        // Should instead refresh tile data in place.
```
No, this was the case with the prior implementation as well, so I'd like to leave it as a followup.

jfirebaugh commented:
I'll write some architecture documentation today.

kkaefer commented on src/mbgl/renderer/style_diff.hpp:
```diff
@@ -11,11 +11,18 @@
+public:
+    T before;
+    T after;
+};
```
:+1:

asheemmamoowala commented:
💯  for the updated Architecture!
